streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala

```diff
@@ -92,29 +95,43 @@ private[streaming] class BlockGenerator(
   @volatile private var stopped = false
 
   /** Start block generating and pushing threads. */
-  def start() {
-    blockIntervalTimer.start()
-    blockPushingThread.start()
-    logInfo("Started BlockGenerator")
+  def start(): Unit = {
+    if (!stopped) {
+      blockIntervalTimer.start()
+      blockPushingThread.start()
+      logInfo("Started BlockGenerator")
+    } else {
+      throw new SparkException("Cannot start BlockGenerator as already stopped")
+    }
   }
 
   /** Stop all threads. */
-  def stop() {
-    logInfo("Stopping BlockGenerator")
-    blockIntervalTimer.stop(interruptTimer = false)
-    stopped = true
-    logInfo("Waiting for block pushing thread")
-    blockPushingThread.join()
-    logInfo("Stopped BlockGenerator")
+  def stop(): Unit = {
+    if (!stopped) {
+      // First stop generating block, then mark it as non-active and wait for pushing thread to stop
+      logInfo("Stopping BlockGenerator")
+      blockIntervalTimer.stop(interruptTimer = false)
+      stopped = true
+      logInfo("Waiting for block pushing thread")
+      blockPushingThread.join()
+      logInfo("Stopped BlockGenerator")
+    } else {
+      logWarning("BlockGenerator not started")
+    }
   }
 
   /**
    * Push a single data item into the buffer. All received data items
    * will be periodically pushed into BlockManager.
    */
-  def addData (data: Any): Unit = synchronized {
-    waitToPush()
-    currentBuffer += data
+  def addData(data: Any): Unit = synchronized {
+    if (!stopped) {
+      waitToPush()
```

Comment:
Do you want to avoid adding any data after BlockGenerator stops? If so, it can still happen because `def stop()` doesn't use `synchronized`.